db/src/db.rs

```diff
@@ -285,60 +283,62 @@ pub fn ensure_current_version(conn: &mut rusqlite::Connection) -> Result<i32> {
-    ValueType::from_value_type_tag(value_type_tag)
-         .ok_or(ErrorKind::BadValueTypeTag(*value_type_tag).into())
-         .and_then(|value_type| -> Result<Value> {
-             match (value_type, value) {
-                 (ValueType::Ref, &rusqlite::types::Value::Integer(ref x)) => Ok(Value::Integer(*x)),
-                 (ValueType::Boolean, &rusqlite::types::Value::Integer(ref x)) => Ok(Value::Boolean(0 != *x)),
-                 (ValueType::Instant, &rusqlite::types::Value::Integer(_)) => bail!(ErrorKind::NotYetImplemented(":db.type/instant".into())),
-                 // SQLite distinguishes integral from decimal types, allowing long and double to
-                 // share a tag.  That means the `value_type` above might be incorrect: we could see
-                 // a `Long` tag and really have `Double` data.  We always trust the tag: Mentat
-                 // should be managing its metadata correctly.
-                 (ValueType::Long, &rusqlite::types::Value::Integer(ref x)) => Ok(Value::Integer(*x)),
-                 (ValueType::Long, &rusqlite::types::Value::Real(ref x)) if value_type_tag == &ValueType::Double.value_type_tag() => Ok(Value::Float((*x).into())),
-                 // This is spurious, since `value_type` never returns `Double`, but let's just
-                 // leave it in case we change things.
-                 (ValueType::Double, &rusqlite::types::Value::Real(ref x)) => Ok(Value::Float((*x).into())),
-                 (ValueType::String, &rusqlite::types::Value::Text(ref x)) => Ok(Value::Text(x.clone())),
-                 (ValueType::UUID, &rusqlite::types::Value::Text(_)) => bail!(ErrorKind::NotYetImplemented(":db.type/uuid".into())),
-                 (ValueType::URI, &rusqlite::types::Value::Text(_)) => bail!(ErrorKind::NotYetImplemented(":db.type/uri".into())),
-                 (ValueType::Keyword, &rusqlite::types::Value::Text(ref x)) => Ok(Value::Text(x.clone())),
-                 (_, value) => bail!(ErrorKind::BadValueAndTagPair(value.clone(), *value_type_tag)),
-             }
-         })
-}
+impl TypedValue {
+    /// Given a SQLite `value` and a `value_type_tag`, return the corresponding `TypedValue`.
+    pub fn from_sql_value_pair(value: &rusqlite::types::Value, value_type_tag: &i32) -> Result<TypedValue> {
+        match (*value_type_tag, value) {
+            (0, &rusqlite::types::Value::Integer(ref x)) => Ok(TypedValue::Ref(*x)),
+            (1, &rusqlite::types::Value::Integer(ref x)) => Ok(TypedValue::Boolean(0 != *x)),
+            // SQLite distinguishes integral from decimal types, allowing long and double to
+            // share a tag.
+            (5, &rusqlite::types::Value::Integer(ref x)) => Ok(TypedValue::Long(*x)),
+            (5, &rusqlite::types::Value::Real(ref x)) => Ok(TypedValue::Double((*x).into())),
+            (10, &rusqlite::types::Value::Text(ref x)) => Ok(TypedValue::String(x.clone())),
```

Comment:
Maybe.  I thought about this and I'm okay with `into_sql_value_pair`; let me try it on for size.